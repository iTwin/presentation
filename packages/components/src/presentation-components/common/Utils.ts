/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Core
 */

import { LegacyRef, MutableRefObject, RefCallback, useCallback, useRef, useState } from "react";
import { Primitives, PrimitiveValue, PropertyDescription, PropertyRecord, PropertyValueFormat } from "@itwin/appui-abstract";
import { IPropertyValueRenderer, PropertyValueRendererManager } from "@itwin/components-react";
import { assert, Guid, GuidString, IDisposable } from "@itwin/core-bentley";
import { Descriptor, Field, LabelCompositeValue, LabelDefinition, parseCombinedFieldNames } from "@itwin/presentation-common";
import { Presentation } from "@itwin/presentation-frontend";
import { InstanceKeyValueRenderer } from "../properties/InstanceKeyValueRenderer";

const localizationNamespaceName = "PresentationComponents";

/**
 * Registers 'PresentationComponents' localization namespace and returns callback
 * to unregister it.
 * @internal
 */
export const initializeLocalization = async () => {
  await Presentation.localization.registerNamespace(localizationNamespaceName);
  return () => Presentation.localization.unregisterNamespace(localizationNamespaceName); // eslint-disable-line @itwin/no-internal
};

/**
 * Registers custom property value renderers and returns cleanup callback that unregisters them.
 * @internal
 */
export const initializePropertyValueRenderers = async () => {
  const customRenderers: Array<{ name: string; renderer: IPropertyValueRenderer }> = [{ name: "SelectableInstance", renderer: new InstanceKeyValueRenderer() }];

  for (const { name, renderer } of customRenderers) {
    PropertyValueRendererManager.defaultManager.registerRenderer(name, renderer);
  }

  return () => {
    for (const { name } of customRenderers) {
      PropertyValueRendererManager.defaultManager.unregisterRenderer(name);
    }
  };
};

/**
 * Translate a string with the specified id from `PresentationComponents`
 * localization namespace. The `stringId` should not contain namespace - it's
 * prepended automatically.
 *
 * @internal
 */
export const translate = (stringId: string): string => {
  stringId = `${localizationNamespaceName}:${stringId}`;
  return Presentation.localization.getLocalizedString(stringId);
};

/**
 * Creates a display name for the supplied component
 * @internal
 */
export const getDisplayName = <P>(component: React.ComponentType<P>): string => {
  if (component.displayName) {
    return component.displayName;
  }
  if (component.name) {
    return component.name;
  }
  return "Component";
};

/**
 * Finds a field given the name of property record created from that field.
 * @internal
 */
export const findField = (descriptor: Descriptor, recordPropertyName: string): Field | undefined => {
  let fieldsSource: { getFieldByName: (name: string) => Field | undefined } | undefined = descriptor;
  const fieldNames = parseCombinedFieldNames(recordPropertyName);
  while (fieldsSource && fieldNames.length) {
    const field: Field | undefined = fieldsSource.getFieldByName(fieldNames.shift()!);
    fieldsSource = field && field.isNestedContentField() ? field : undefined;
    if (!fieldNames.length) {
      return field;
    }
  }
  return undefined;
};

/**
 * Creates property record for label using label definition.
 * @internal
 */
export const createLabelRecord = (label: LabelDefinition, name: string): PropertyRecord => {
  const value: PrimitiveValue = {
    displayValue: label.displayValue,
    value: createPrimitiveLabelValue(label),
    valueFormat: PropertyValueFormat.Primitive,
  };
  const property: PropertyDescription = {
    displayLabel: "Label",
    typename: label.typeName,
    name,
  };

  return new PropertyRecord(value, property);
};

const createPrimitiveLabelValue = (label: LabelDefinition) => {
  return LabelDefinition.isCompositeDefinition(label) ? createPrimitiveCompositeValue(label.rawValue) : label.rawValue;
};

const createPrimitiveCompositeValue = (compositeValue: LabelCompositeValue): Primitives.Composite => {
  return {
    separator: compositeValue.separator,
    parts: compositeValue.values.map((part) => ({
      displayValue: part.displayValue,
      typeName: part.typeName,
      rawValue: createPrimitiveLabelValue(part),
    })),
  };
};

/**
 * A helper to track ongoing async tasks. Usage:
 * ```
 * await using(tracker.trackAsyncTask(), async (_r) => {
 *   await doSomethingAsync();
 * });
 * ```
 *
 * Can be used with `waitForPendingAsyncs` in test helpers to wait for all
 * async tasks to complete.
 *
 * @internal
 */
export class AsyncTasksTracker {
  private _asyncsInProgress = new Set<GuidString>();
  public get pendingAsyncs() {
    return this._asyncsInProgress;
  }
  public trackAsyncTask(): IDisposable {
    const id = Guid.createValue();
    this._asyncsInProgress.add(id);
    return {
      dispose: () => this._asyncsInProgress.delete(id),
    };
  }
}

/** @internal */
export function useResizeObserver<T extends HTMLElement>() {
  const observer = useRef<ResizeObserver>();
  const [{ width, height }, setSize] = useState<{ width?: number; height?: number }>({});

  const ref = useCallback((element: T | null) => {
    observer.current?.disconnect();
    if (element) {
      observer.current = new ResizeObserver(
        /* istanbul ignore next */
        (entries) => {
          assert(entries.length === 1);
          setSize(entries[0].contentRect);
        }
      );
      observer.current.observe(element);
    }
  }, []);

  return {
    ref,
    width,
    height,
  };
}

/** @internal */
export function mergeRefs<T>(...refs: Array<MutableRefObject<T | null> | LegacyRef<T>>): RefCallback<T> {
  return (instance: T | null) => {
    refs.forEach((ref) => {
      // istanbul ignore else
      if (typeof ref === "function") {
        ref(instance);
      } else if (ref) {
        (ref as MutableRefObject<T | null>).current = instance;
      }
    });
  };
}

/**
 * A hook that helps components throw errors in React's render loop so they can be captured by React error
 * boundaries.
 *
 * Usage: simply call the returned function with an error and it will be re-thrown on next render.
 *
 * @internal
 */
export function useErrorState() {
  const [error, setError] = useState<Error | undefined>(undefined);
  const setErrorState = useCallback((e: unknown) => {
    setError(e instanceof Error ? e : /* istanbul ignore next */ new Error());
  }, []);
  if (error) {
    throw error;
  }
  return setErrorState;
}
