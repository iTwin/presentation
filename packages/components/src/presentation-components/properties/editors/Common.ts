/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/

import { PropertyValue } from "@itwin/appui-abstract";

/** @internal */
export interface PropertyEditorAttributes {
  getValue: () => PropertyValue | undefined;
  htmlElement: HTMLElement | null;
}
