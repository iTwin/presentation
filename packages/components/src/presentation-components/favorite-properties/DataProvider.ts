/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module FavoriteProperties
 */

import { PropertyData } from "@itwin/components-react";
import { Id64Arg } from "@itwin/core-bentley";
import { IModelConnection } from "@itwin/core-frontend";
import { KeySet, Ruleset } from "@itwin/presentation-common";
import { createECSqlQueryExecutor } from "@itwin/presentation-core-interop";
import { Presentation } from "@itwin/presentation-frontend";
import { computeSelection } from "@itwin/unified-selection";
import { mapPresentationFrontendSelectionScopeToUnifiedSelectionScope } from "../common/Utils.js";
import { PresentationPropertyDataProvider } from "../propertygrid/DataProvider.js";
import { getFavoritesCategory } from "./Utils.js";

/**
 * An data provider interface for returning favorite properties for the given elements
 * @public
 */
export interface IFavoritePropertiesDataProvider {
  /** Returns property data for an element. */
  getData: (imodel: IModelConnection, elementIds: Id64Arg | KeySet) => Promise<PropertyData>;
}

/**
 * Props for [[FavoritePropertiesDataProvider]]
 * @public
 */
export interface FavoritePropertiesDataProviderProps {
  /**
   * Id of the ruleset to use when requesting properties or a ruleset itself. If not
   * set, default presentation rules are used which return content for the selected elements.
   */
  ruleset?: Ruleset | string;

  /**
   * Active selection scope provider.
   * Takes active scope from `Presentation.selection.scopes.activeScope` if not provided.
   */
  activeScopeProvider?: () => Parameters<typeof computeSelection>[0]["scope"];
}

/**
 * Presentation Rules-driven element favorite properties data provider implementation.
 * @public
 */
export class FavoritePropertiesDataProvider implements IFavoritePropertiesDataProvider {
  private _customRuleset?: Ruleset | string;
  private _getActiveScope: () => Parameters<typeof computeSelection>[0]["scope"];

  /**
   * Should fields with no values be included in the property list. No value means:
   * - For *primitive* fields: `null`, `undefined`, `""` (empty string)
   * - For *array* fields: `[]` (empty array)
   * - For *struct* fields: `{}` (object with no members)
   */
  public includeFieldsWithNoValues: boolean;

  /**
   * Should fields with composite values be included in the property list.
   * Fields with composite values:
   * - *array* fields.
   * - *struct* fields.
   */
  public includeFieldsWithCompositeValues: boolean;

  /** Constructor. */
  constructor(props?: FavoritePropertiesDataProviderProps) {
    this.includeFieldsWithNoValues = true;
    this.includeFieldsWithCompositeValues = true;
    this._customRuleset = /* c8 ignore next */ props?.ruleset;
    this._getActiveScope =
      props?.activeScopeProvider ?? (() => mapPresentationFrontendSelectionScopeToUnifiedSelectionScope(Presentation.selection.scopes.activeScope));
  }

  /**
   * Returns PropertyData for the specified elements.
   * PropertyData only contains a single category for favorite properties (if there are any).
   */
  public async getData(imodel: IModelConnection, elementIds: Id64Arg | KeySet): Promise<PropertyData> {
    if (elementIds instanceof KeySet) {
      using propertyDataProvider = this.createPropertyDataProvider(imodel, this._customRuleset);
      propertyDataProvider.keys = elementIds;
      // eslint-disable-next-line @typescript-eslint/no-deprecated
      propertyDataProvider.includeFieldsWithNoValues = this.includeFieldsWithNoValues;
      // eslint-disable-next-line @typescript-eslint/no-deprecated
      propertyDataProvider.includeFieldsWithCompositeValues = this.includeFieldsWithCompositeValues;
      const propertyData = await propertyDataProvider.getData();

      // leave only favorite properties
      const favoritesCategory = getFavoritesCategory();
      propertyData.categories = propertyData.categories.filter((c) => c.name === favoritesCategory.name);
      propertyData.records = propertyData.records.hasOwnProperty(favoritesCategory.name)
        ? { [favoritesCategory.name]: propertyData.records[favoritesCategory.name] }
        : {};
      return propertyData;
    }

    const iter = computeSelection({
      queryExecutor: createECSqlQueryExecutor(imodel),
      scope: this._getActiveScope(),
      elementIds,
    });
    const keys = new KeySet();
    for await (const key of iter) {
      keys.add(key);
    }
    return this.getData(imodel, keys);
  }

  /* c8 ignore start */
  private createPropertyDataProvider(imodel: IModelConnection, ruleset?: Ruleset | string) {
    const provider = new PresentationPropertyDataProvider({ imodel, ruleset });
    provider.isNestedPropertyCategoryGroupingEnabled = false;
    return provider;
  }
  /* c8 ignore end */
}
