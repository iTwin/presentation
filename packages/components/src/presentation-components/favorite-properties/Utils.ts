/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/

import { CategoryDescription } from "@itwin/presentation-common";
import { translate } from "../common/Utils";

/** @internal */
export const FAVORITES_CATEGORY_NAME = "Favorite";

/** @internal */
export const getFavoritesCategory = (): CategoryDescription => {
  return {
    name: FAVORITES_CATEGORY_NAME,
    label: translate("categories.favorite.label"),
    description: translate("categories.favorite.description"),
    priority: Number.MAX_VALUE,
    expand: true,
  };
};
