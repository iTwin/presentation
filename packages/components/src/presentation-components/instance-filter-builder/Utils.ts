/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module InstancesFilter
 */

import { useMemo } from "react";
import { PrimitiveValue, PropertyDescription, PropertyValueFormat, StandardTypeNames } from "@itwin/appui-abstract";
import {
  defaultPropertyFilterBuilderRuleValidator,
  isUnaryPropertyFilterOperator,
  PropertyFilterBuilderRule,
  PropertyFilterRuleOperator,
} from "@itwin/components-react";
import { IModelConnection } from "@itwin/core-frontend";
import { CategoryDescription, ClassInfo, combineFieldNames, Descriptor, Field, NestedContentField, PropertiesField } from "@itwin/presentation-common";
import { createPropertyDescriptionFromFieldInfo } from "../common/ContentBuilder";
import { translate } from "../common/Utils";
import { NavigationPropertyEditorContextProps } from "../properties/editors/NavigationPropertyEditorContext";
import { PresentationInstanceFilterPropertyInfo } from "./PresentationFilterBuilder";

/** @internal */
export function createInstanceFilterPropertyInfos(descriptor: Descriptor): PresentationInstanceFilterPropertyInfo[] {
  const propertyInfos = new Array<PresentationInstanceFilterPropertyInfo>();
  for (const field of descriptor.fields) {
    propertyInfos.push(...createPropertyInfos(field));
  }
  return propertyInfos;
}

/** @internal */
export function getInstanceFilterFieldName(property: PropertyDescription) {
  const [_, fieldName] = property.name.split(INSTANCE_FILTER_FIELD_SEPARATOR);
  return fieldName;
}

/** @internal */
export const DEFAULT_ROOT_CATEGORY_NAME = "/selected-item/";

function getPropertySourceClassInfo(field: PropertiesField | NestedContentField): ClassInfo {
  if (field.parent) {
    return getPropertySourceClassInfo(field.parent);
  }

  if (field.isPropertiesField()) {
    return field.properties[0].property.classInfo;
  }
  return field.pathToPrimaryClass[field.pathToPrimaryClass.length - 1].targetClassInfo;
}

function getPropertyClassInfo(field: PropertiesField): ClassInfo {
  if (field.parent && field.parent.isNestedContentField()) {
    return field.parent.contentClassInfo;
  }

  return field.properties[0].property.classInfo;
}

function createPropertyInfos(field: Field): PresentationInstanceFilterPropertyInfo[] {
  if (field.isNestedContentField()) {
    return field.nestedFields.flatMap((nestedField) => createPropertyInfos(nestedField));
  }
  // istanbul ignore else
  if (field.isPropertiesField()) {
    return [createPropertyInfoFromPropertiesField(field)];
  }
  // istanbul ignore next
  return [];
}

interface CategoryInfo {
  name?: string;
  label?: string;
}

function getCategoryInfo(category: CategoryDescription, categoryInfo: CategoryInfo): CategoryInfo {
  if (category.name === DEFAULT_ROOT_CATEGORY_NAME) {
    return categoryInfo;
  }

  const newInfo = {
    name: categoryInfo.name ? `${category.name}/${categoryInfo.name}` : `${category.name}`,
    label: categoryInfo.label ? `${category.label} | ${categoryInfo.label}` : `${category.label}`,
  };

  return category.parent ? getCategoryInfo(category.parent, newInfo) : newInfo;
}

function getParentNames(field: Field, name: string): string {
  if (!field.parent) {
    return combineFieldNames(name, field.name);
  }
  return getParentNames(field.parent, combineFieldNames(name, field.name));
}

/** @internal */
export function createPropertyInfoFromPropertiesField(field: PropertiesField): PresentationInstanceFilterPropertyInfo {
  const categoryInfo = getCategoryInfo(field.category, { name: undefined, label: undefined });
  const name = field.parent ? getParentNames(field.parent, field.name) : field.name;

  const propertyDescription = createPropertyDescriptionFromFieldInfo({
    name: getCategorizedFieldName(name, categoryInfo.name),
    label: field.label,
    type: field.type,
    editor: field.editor,
    enum: field.properties[0].property.enumerationInfo,
    isReadonly: field.isReadonly,
    renderer: field.renderer,
    koqName: field.properties[0].property.kindOfQuantity?.name,
  });

  return {
    field,
    sourceClassId: getPropertySourceClassInfo(field).id,
    propertyDescription,
    categoryLabel: categoryInfo.label,
    className: getPropertyClassInfo(field).name,
  };
}

/** @internal */
export const INSTANCE_FILTER_FIELD_SEPARATOR = "#";
function getCategorizedFieldName(fieldName: string, categoryName?: string) {
  return `${categoryName ?? ""}${INSTANCE_FILTER_FIELD_SEPARATOR}${fieldName}`;
}

/** @internal */
export function useFilterBuilderNavigationPropertyEditorContext(imodel: IModelConnection, descriptor: Descriptor) {
  return useMemo<NavigationPropertyEditorContextProps>(
    () => ({
      imodel,
      getNavigationPropertyInfo: async (property) => {
        const field = descriptor.getFieldByName(getInstanceFilterFieldName(property));
        if (!field || !field.isPropertiesField()) {
          return undefined;
        }

        return field.properties[0].property.navigationPropertyInfo;
      },
    }),
    [imodel, descriptor],
  );
}

/** @internal */
export function filterRuleValidator(item: PropertyFilterBuilderRule) {
  // skip empty rules and rules that do not require value
  if (item.property === undefined || item.operator === undefined || isUnaryPropertyFilterOperator(item.operator)) {
    return undefined;
  }

  // istanbul ignore if
  if (item.value !== undefined && item.value.valueFormat !== PropertyValueFormat.Primitive) {
    return undefined;
  }

  const error = combineValidators(
    quantityPropertyValidator,
    numericPropertyValidator,
  )({
    property: item.property,
    operator: item.operator,
    value: item.value,
  });

  if (error) {
    return error;
  }

  return defaultPropertyFilterBuilderRuleValidator(item);
}

function combineValidators(...validators: Array<(ctx: ValidatorContext) => string | undefined>) {
  return (ctx: ValidatorContext) => {
    for (const validator of validators) {
      const error = validator(ctx);
      if (error) {
        return error;
      }
    }
    return undefined;
  };
}

interface ValidatorContext {
  property: PropertyDescription;
  operator: PropertyFilterRuleOperator;
  value?: PrimitiveValue;
}

function quantityPropertyValidator({ property, value }: ValidatorContext) {
  // rules with non quantity properties or without values
  if (property.quantityType === undefined || value === undefined) {
    return undefined;
  }
  if (isInvalidNumericValue(value)) {
    return translate("instance-filter-builder.error-messages.invalid");
  }

  return undefined;
}

function numericPropertyValidator({ property, value }: ValidatorContext) {
  if (!isPropertyNumeric(property.typename) || value === undefined) {
    return undefined;
  }

  if (isInvalidNumericValue(value)) {
    return translate("instance-filter-builder.error-messages.not-a-number");
  }

  return undefined;
}

function isPropertyNumeric(typename: string) {
  return (
    typename === StandardTypeNames.Number || typename === StandardTypeNames.Int || typename === StandardTypeNames.Float || typename === StandardTypeNames.Double
  );
}

function isInvalidNumericValue(value: PrimitiveValue) {
  return value.displayValue !== undefined && value.displayValue !== "" && isNaN(Number(value.value));
}
