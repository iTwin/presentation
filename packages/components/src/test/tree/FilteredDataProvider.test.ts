/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/

import { expect } from "chai";
import sinon from "sinon";
import * as moq from "typemoq";
import { PageOptions } from "@itwin/components-react";
import { BeEvent } from "@itwin/core-bentley";
import { IModelConnection } from "@itwin/core-frontend";
import { LabelDefinition, NodePathElement } from "@itwin/presentation-common";
import { Presentation, PresentationManager, RulesetVariablesManager } from "@itwin/presentation-frontend";
import { PresentationTreeDataProvider } from "../../presentation-components/tree/DataProvider";
import { FilteredPresentationTreeDataProvider } from "../../presentation-components/tree/FilteredDataProvider";
import { IPresentationTreeDataProvider } from "../../presentation-components/tree/IPresentationTreeDataProvider";
import { createTreeNodeItem } from "../../presentation-components/tree/Utils";
import { createTestECInstanceKey } from "../_helpers/Common";
import { createTestECInstancesNode, createTestECInstancesNodeKey, createTestNodePathElement } from "../_helpers/Hierarchy";
import { createTestTreeNodeItem } from "../_helpers/UiComponents";

describe("FilteredTreeDataProvider", () => {
  function createTestNodePathElementWithId(id: string) {
    return createTestNodePathElement({
      node: createTestECInstancesNode({
        key: createTestECInstancesNodeKey({
          instanceKeys: [createTestECInstanceKey({ id })],
          pathFromRoot: [id],
        }),
      }),
    });
  }

  function createPaths() {
    /*
    A-1
      A-1-1
    A-2
      A-2-1
      A-2-2
        A-2-2-1
    */
    const nodePaths: NodePathElement[] = [];

    nodePaths[0] = createTestNodePathElementWithId("0x1");
    nodePaths[0].node.label = LabelDefinition.fromLabelString("A-1");

    nodePaths[1] = createTestNodePathElementWithId("0x2");
    nodePaths[1].node.label = LabelDefinition.fromLabelString("A-2");

    nodePaths[0].children = [];
    nodePaths[0].children[0] = createTestNodePathElementWithId("0x3");
    nodePaths[0].children[0].node.label = LabelDefinition.fromLabelString("A-1-1");

    nodePaths[1].children = [];
    nodePaths[1].children[0] = createTestNodePathElementWithId("0x4");
    nodePaths[1].children[0].node.label = LabelDefinition.fromLabelString("A-2-1");

    nodePaths[1].children[1] = createTestNodePathElementWithId("0x5");
    nodePaths[1].children[1].node.label = LabelDefinition.fromLabelString("A-2-2");

    nodePaths[1].children[1].children = [];
    nodePaths[1].children[1].children[0] = createTestNodePathElementWithId("0x6");
    nodePaths[1].children[1].children[0].node.label = LabelDefinition.fromLabelString("A-2-2-1");
    return nodePaths;
  }

  let provider: FilteredPresentationTreeDataProvider;
  let filter: string;
  let paths: NodePathElement[];
  const parentProviderMock = moq.Mock.ofType<IPresentationTreeDataProvider>();
  const imodelMock = moq.Mock.ofType<IModelConnection>();
  const pageOptions: PageOptions = { size: 0, start: 0 };

  beforeEach(() => {
    const onVariableChanged = new BeEvent();
    const presentationManagerMock = moq.Mock.ofType<PresentationManager>();
    const rulesetVariablesManagerMock = moq.Mock.ofType<RulesetVariablesManager>();
    presentationManagerMock.setup((x) => x.vars(moq.It.isAny())).returns(() => rulesetVariablesManagerMock.object);
    rulesetVariablesManagerMock.setup((x) => x.onVariableChanged).returns(() => onVariableChanged);
    sinon.stub(Presentation, "presentation").get(() => presentationManagerMock.object);

    parentProviderMock.reset();
    filter = "test_filter";
    paths = createPaths();
    provider = new FilteredPresentationTreeDataProvider({
      parentDataProvider: parentProviderMock.object,
      filter,
      paths,
    });
  });

  afterEach(() => {
    sinon.restore();
  });

  describe("filter", () => {
    it("returns filter with which it was initialized", () => {
      expect(provider.filter).to.be.equal(filter);
    });
  });

  describe("rulesetId", () => {
    it("returns rulesetId of the parent data provider", () => {
      const expectedRulesetId = "ruleset_id";
      parentProviderMock
        .setup((x) => x.rulesetId)
        .returns(() => expectedRulesetId)
        .verifiable();
      expect(provider.rulesetId).to.eq(expectedRulesetId);
      parentProviderMock.verifyAll();
    });
  });

  describe("imodel", () => {
    it("returns imodel of the parent data provider", () => {
      parentProviderMock
        .setup((x) => x.imodel)
        .returns(() => imodelMock.object)
        .verifiable();
      expect(provider.imodel).to.eq(imodelMock.object);
      parentProviderMock.verifyAll();
    });
  });

  describe("parentDataProvider", () => {
    it("returns parent data provider", () => {
      expect(provider.parentDataProvider).to.eq(parentProviderMock.object);
    });
  });

  describe("getNodes", () => {
    it("returns root nodes", async () => {
      const result = await provider.getNodes(undefined, pageOptions);
      expect(result).to.matchSnapshot();
    });

    it("returns child nodes", async () => {
      const parentNode = createTreeNodeItem(paths[1].node);

      const result = await provider.getNodes(parentNode, pageOptions);
      expect(result).to.matchSnapshot();
    });

    it("applies same node customizations as parent data provider", async () => {
      const customizeStub = sinon.stub();
      const parentProvider = new PresentationTreeDataProvider({
        imodel: imodelMock.object,
        ruleset: "test-rules",
        customizeTreeNodeItem: customizeStub,
      });
      const testPaths = [createTestNodePathElement()];
      const filteredProvider = new FilteredPresentationTreeDataProvider({
        parentDataProvider: parentProvider,
        filter: "Test",
        paths: testPaths,
      });
      await filteredProvider.getNodes(undefined, pageOptions);
      expect(customizeStub).to.be.calledOnce;
    });
  });

  describe("getNodesCount", () => {
    it("returns root nodes count", async () => {
      const result = await provider.getNodesCount();
      expect(result).to.equal(paths.length);
    });

    it("returns child nodes count", async () => {
      const parentNode = createTreeNodeItem(paths[1].node);

      const result = await provider.getNodesCount(parentNode);
      expect(result).to.equal(paths[1].children.length);
    });
  });

  describe("getFilteredNodePaths", () => {
    it("calls parent data provider", async () => {
      parentProviderMock
        .setup(async (x) => x.getFilteredNodePaths(filter))
        .returns(async () => paths)
        .verifiable();

      const result = await provider.getFilteredNodePaths(filter);
      expect(result).to.equal(paths);
      parentProviderMock.verifyAll();
    });
  });

  describe("getNodeKey", () => {
    it("returns node key", () => {
      const key = createTestECInstancesNodeKey();
      const treeNode = createTestTreeNodeItem(key);

      parentProviderMock.setup((x) => x.getNodeKey(treeNode)).returns(() => key); // eslint-disable-line deprecation/deprecation
      const result = provider.getNodeKey(treeNode); // eslint-disable-line deprecation/deprecation
      expect(result).to.deep.equal(key);
    });
  });

  const constantFilter = "test";
  const filteredNodePaths: NodePathElement[] = [];

  filteredNodePaths[0] = createTestNodePathElementWithId("0x1");
  filteredNodePaths[0].node.label = LabelDefinition.fromLabelString("A-1");
  filteredNodePaths[0].filteringData = { matchesCount: 0, childMatchesCount: 1 };

  filteredNodePaths[0].children = [];
  filteredNodePaths[0].children[0] = createTestNodePathElementWithId("0x2");
  filteredNodePaths[0].children[0].node.label = LabelDefinition.fromLabelString("A-1-1 test");
  filteredNodePaths[0].children[0].filteringData = { matchesCount: 1, childMatchesCount: 0 };

  filteredNodePaths[1] = createTestNodePathElementWithId("0x3");
  filteredNodePaths[1].node.label = LabelDefinition.fromLabelString("A-2 test");
  filteredNodePaths[1].filteringData = { matchesCount: 1, childMatchesCount: 0 };

  filteredNodePaths[1].children = [];
  filteredNodePaths[1].children[0] = createTestNodePathElementWithId("0x4");
  filteredNodePaths[1].children[0].node.label = LabelDefinition.fromLabelString("A-2-1");
  filteredNodePaths[1].children[0].filteringData = { matchesCount: 0, childMatchesCount: 0 };

  filteredNodePaths[1].children[1] = createTestNodePathElementWithId("0x5");
  filteredNodePaths[1].children[1].node.label = LabelDefinition.fromLabelString("A-2-2");
  filteredNodePaths[1].children[1].filteringData = { matchesCount: 0, childMatchesCount: 0 };

  describe("countFilteringResults", () => {
    it("all matches get counted", () => {
      expect(provider.countFilteringResults(filteredNodePaths)).to.be.eq(2);
    });

    it("doesn't count if node paths don't have filtering data", () => {
      paths = [];
      paths[0] = createTestNodePathElement();
      paths[0].node.label = LabelDefinition.fromLabelString("A-1");
      paths[0].filteringData = undefined;
      expect(provider.countFilteringResults(paths)).to.eq(0);
    });
  });

  describe("getActiveMatch", () => {
    it("returns correct match", () => {
      provider = new FilteredPresentationTreeDataProvider({
        parentDataProvider: parentProviderMock.object,
        filter: constantFilter,
        paths: filteredNodePaths,
      });
      const result = provider.getActiveMatch(2);

      expect(result).to.not.be.undefined;
      expect(result!.nodeId).to.be.eq(createTreeNodeItem(filteredNodePaths[1].node).id);
      expect(result!.matchIndex).to.be.eq(0);
    });

    it("returns undefined when index is 0 or lower", () => {
      provider = new FilteredPresentationTreeDataProvider({
        parentDataProvider: parentProviderMock.object,
        filter: constantFilter,
        paths: filteredNodePaths,
      });
      const result = provider.getActiveMatch(0);
      expect(result).to.be.undefined;
    });
  });

  describe("nodeMatchesFilter", () => {
    it("returns true when node matches filter", () => {
      provider = new FilteredPresentationTreeDataProvider({
        parentDataProvider: parentProviderMock.object,
        filter: constantFilter,
        paths: filteredNodePaths,
      });
      const node = createTreeNodeItem(filteredNodePaths[1].node);
      expect(provider.nodeMatchesFilter(node)).to.be.true;
    });

    it("returns false when node matches filter", () => {
      provider = new FilteredPresentationTreeDataProvider({
        parentDataProvider: parentProviderMock.object,
        filter: constantFilter,
        paths: filteredNodePaths,
      });
      const node = createTreeNodeItem(filteredNodePaths[0].node);
      expect(provider.nodeMatchesFilter(node)).to.be.false;
    });
  });
});
