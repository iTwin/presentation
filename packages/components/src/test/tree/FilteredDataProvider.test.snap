// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilteredTreeDataProvider getNodes returns child nodes 1`] = `
Array [
  Object {
    "__presentation-components/key": Object {
      "instanceKeys": Array [
        Object {
          "className": "TestSchema:TestClass",
          "id": "0x4",
        },
      ],
      "pathFromRoot": Array [
        "0x4",
      ],
      "type": "ECInstancesNode",
      "version": 2,
    },
    "description": "testDescription",
    "id": "0x4",
    "label": PropertyRecord {
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "A-2-1",
        "value": "A-2-1",
        "valueFormat": 0,
      },
    },
    "parentId": "0x2",
  },
  Object {
    "__presentation-components/key": Object {
      "instanceKeys": Array [
        Object {
          "className": "TestSchema:TestClass",
          "id": "0x5",
        },
      ],
      "pathFromRoot": Array [
        "0x5",
      ],
      "type": "ECInstancesNode",
      "version": 2,
    },
    "autoExpand": true,
    "description": "testDescription",
    "hasChildren": true,
    "id": "0x5",
    "label": PropertyRecord {
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "A-2-2",
        "value": "A-2-2",
        "valueFormat": 0,
      },
    },
    "parentId": "0x2",
  },
]
`;

exports[`FilteredTreeDataProvider getNodes returns root nodes 1`] = `
Array [
  Object {
    "__presentation-components/key": Object {
      "instanceKeys": Array [
        Object {
          "className": "TestSchema:TestClass",
          "id": "0x1",
        },
      ],
      "pathFromRoot": Array [
        "0x1",
      ],
      "type": "ECInstancesNode",
      "version": 2,
    },
    "autoExpand": true,
    "description": "testDescription",
    "hasChildren": true,
    "id": "0x1",
    "label": PropertyRecord {
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "A-1",
        "value": "A-1",
        "valueFormat": 0,
      },
    },
  },
  Object {
    "__presentation-components/key": Object {
      "instanceKeys": Array [
        Object {
          "className": "TestSchema:TestClass",
          "id": "0x2",
        },
      ],
      "pathFromRoot": Array [
        "0x2",
      ],
      "type": "ECInstancesNode",
      "version": 2,
    },
    "autoExpand": true,
    "description": "testDescription",
    "hasChildren": true,
    "id": "0x2",
    "label": PropertyRecord {
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "A-2",
        "value": "A-2",
        "valueFormat": 0,
      },
    },
  },
]
`;
